chalk-solve/src/clauses.rs

```diff
@@ -262,8 +261,19 @@ fn program_clauses_that_could_match<I: Interner>(
                         let qwc =
                             exists_qwc.substitute(interner, &[self_ty.clone().cast(interner)]);
 
-                        builder.push_binders(&qwc, |builder, wc| {
-                            builder.push_fact(wc);
+                        builder.push_binders(&qwc, |builder, wc| match &wc {
+                            // For the implemented traits, we need to elaborate super traits and add where clauses from the trait
+                            WhereClause::Implemented(trait_ref) => {
```

Comment:
Hmm, I wonder if the actual flaw here is that for a type `dyn Foo`, we generate the fact `Implemented(dyn Foo: Foo)` instead of `FromEnv(dyn Foo: Foo)`. The latter would -- through implied bounds -- mean that `FromEnv(dyn Foo: Bar)` and so forth would come for free.

Although, I'm remembering now that there were some subtle correctness conditions to consider here -- @scalexm outlined a proposal in https://github.com/rust-lang/chalk/issues/203 that aimed to address exactly this. 

Regardless, I think that the approach in this PR is working harder than we exactly want.

Reply:
Well yeah, scalexm's rules from #203 are what I tried to implement here :sweat_smile:

Reply:
Using `FromEnv` would only work if we then do env elaboration on it, right?

Reply:
Maybe I'm missing something. I think that if we generate the rule

```
FromEnv(dyn Foo: Foo) // rule 1
```

for any `dyn`-self type, then this would combine with the rules we genreate for the trait `Foo`:

```
// given `trait Foo: Bar`...
Implemented(?T: Foo) :- FromEnv(?T: Foo) // rule 2
FromEnv(?T: Bar) :- FromEnv(?T: Foo) // rule 3
```

such that we can prove both

```
Implemented(dyn Foo: Foo) // because rule 1 and then rule 2
```

and

```
Implemented(dyn Foo: Bar) // because rule 1, rule 3, and then rule 2
```

Reply:
However, I better re-read #203 in full, because I am concerned this would also allow us to prove too much, and I forget exactly what was in there. :)

Reply:
Oh, I think I misunderstood what role env elaboration plays, `FromEnv` would probably work then. I could try it.

But yeah, #203 suggests more complicated rules, so I thought it couldn't be that simple :wink:

Reply:
In particular I think a few is something like

```rust
trait Foo<T: Copy> { ... }
```

where you might have a type like `dyn Foo<Box<u32>>`, and you want to be sure that just *seeing* a type like this isn't going to somehow let you start proving things like `Box<u32>: Copy`. 

The question kind of comes back to "who is ultimately proving that type is well-formed, and does that proof require truly proving `Box<u32>: Copy` (which would fail).

Reply:
OK, re-reading the issue, I think that this is indeed probably what @scalexm meant. The rule I proposed (adding a `FromEnv` into the environment for any `dyn Foo: Foo` goal...) doesn't really make sense. The whole point of `FromEnv` is that it's supposed to be matched with some `WellFormed` obligation proved somewhere else, and in this case -- who is proving it? Nobody.